### ISSUE TITLE
Gateway proxy doesn't work

### ISSUE THREAD

**Issue by huyz-git** [springzfx/cgproxy#1]
PC use LAN for network, and create wireless AP for other device.
If stop `cgproxy.service`, then AP works fine. Other device connected to the AP can access internet.

However, if start `cgproxy.service`, then whatever enable `enable_gateway` or not, other device connected to the AP cannot access internet anymore.

Whatever enable `enable_gateway` or not, PC's transparent proxy works without any problem.

Perhaps it's a DNS issue, because if I run `dig` command on other device connected to the AP, I got the following error whatever enable `enable_gateway` or not (this one runs on an Android phone, using termux):
```
dig baidu.com
;; reply from unexpected source: 10.42.0.1#42050, expected 8.8.8.8#53

;; reply from unexpected source: 10.42.0.1#42605, expected 8.8.4.4#53
```
Change Android phone's DNS to some public DNS has no effect. The error is the same.

Here's the `iptables` rule after enable `enable_gateway` and start `cgproxy.service`:
```
$ iptables -t mangle -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
TPROXY_PRE  all  --  anywhere             anywhere            
CONNMARK   all  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL ctstate NEW CONNMARK set 0x2

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
TPROXY_OUT  all  --  anywhere             anywhere            

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain TPROXY_OUT (1 references)
target     prot opt source               destination         
RETURN     all  --  anywhere             anywhere            
RETURN     icmp --  anywhere             anywhere            
RETURN     all  --  anywhere             anywhere             connmark match  0x2
RETURN     all  --  anywhere             anywhere             mark match 0xff
RETURN     all  --  anywhere             anywhere             cgroup /noproxy.slice
MARK       all  --  anywhere             anywhere             cgroup / MARK set 0x1

Chain TPROXY_PRE (1 references)
target     prot opt source               destination         
RETURN     all  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL
RETURN     all  --  anywhere             anywhere             PKTTYPE = broadcast
RETURN     all  --  anywhere             anywhere             PKTTYPE = multicast
TPROXY     tcp  --  anywhere             anywhere             TPROXY redirect 127.0.0.1:12345 mark 0x1/0xffffffff
TPROXY     udp  --  anywhere             anywhere             TPROXY redirect 127.0.0.1:12345 mark 0x1/0xffffffff
```
And here's the ip route and ip rule:
```
$ ip rule
0:      from all lookup local
32765:  from all fwmark 0x1 lookup 100
32766:  from all lookup main
32767:  from all lookup default

$ ip route
default via 192.168.0.1 dev enp114s0 proto static metric 100 
10.42.0.0/24 dev wlp116s0 proto kernel scope link src 10.42.0.1 metric 600 
192.168.0.0/24 dev enp114s0 proto kernel scope link src 192.168.0.170 metric 100 
192.168.59.0/24 dev vmnet8 proto kernel scope link src 192.168.59.1 
192.168.203.0/24 dev vmnet1 proto kernel scope link src 192.168.203.1 
```

**Comment by springzfx** [comment]
I thinks your proxy program need some special permission, for exmple
`sudo setcap "cap_net_admin,cap_net_bind_service=ep" /usr/lib/v2ray/v2ray`

**Comment by huyz-git** [comment]
> I thinks your proxy program need some special permission, for exmple
> `sudo setcap "cap_net_admin,cap_net_bind_service=ep" /usr/lib/v2ray/v2ray`

I have already set these permissions:
```
$ getcap /usr/lib/v2ray/v2ray 
/usr/lib/v2ray/v2ray = cap_net_bind_service,cap_net_admin+ep
```
Without these two permissions, even PC's transparent proxy doesn't work (has similar error like that).
After set these two permissions, PC's transparent proxy works, but other devices connected to AP still doesn't work.
I also tried to run `v2ray` as root, however the result is the same: PC works but AP doesn't work. The error is exactly the same.

**Comment by springzfx** [comment]
OK, have you tried run  `cgnoproxy <PROXY PROGRAM>` ?  
I have no problem in this way.

**Comment by huyz-git** [comment]
> OK, have you tried run `cgnoproxy <PROXY PROGRAM>` ?
> I have no problem in this way.

Yes, on PC, `cgnoproxy` also works without any problem.
On other device there're no `cgnoproxy` command.

I do some trying, and find out that perhaps the route from PC to phone is incorrect.
If my AP's IP is 10.x.x.x, and if I add this line:
```
iptables -t mangle -A TPROXY_OUT -d 10.0.0.0/8 -j RETURN
```
Then my phone can access internet, and can use PC's transparent proxy.

**Comment by springzfx** [comment]
You are right. I thought conntrack wiil handle this, the fact is not.
I will make a update to fix this.
and btw `cgnoproxy <v2ray ...>` only need to run on your PC.

**Comment by springzfx** [comment]
fixed in 168779ea9003c75577eaedde16ba4cc6d1087dd4
welcome to test it.
